### ISSUE TITLE
Missing coverage on Exception to K&R style blocks

### ISSUE THREAD

**Issue by mohitsatr** [checkstyle/checkstyle#17569]
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

From: https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style
> Exception: In places where these rules allow a single statement ending with a semicolon (;), a block of statements can appear, and the opening brace of this block is preceded by a line break. Blocks like these are typically introduced to limit the scope of local variables.

It means when a block is used to create a smaller variable scope, we can skip formatting it K&R style

Code:
```
/** some javadoc. */
public class Test {

  void test1(int[] nums) {
    if (nums.length > 2)
    { // false-positive
      int temp = nums[0];
      use(temp);
    } // false-positive

    for (int i = 1; i < nums.length; i++) 
    {  // false-positive
      int k = i * 2;
      use(k);
    } // false-positive
  }

  void test2(int k) {
    {  // ok
      int x = k + 2;
      use(x);
    }
  }

  void use(int n) {
    n++;
  }
}
```

Cli:
```
$ java -jar checkstyle-10.26.1-all.jar -c google_checks.xml Test.java 
Starting audit...
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:6:5: 'if lcurly' has incorrect indentation level 4, expected level should be 6. [Indentation]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:6:5: '{' at column 5 should be on the previous line. [LeftCurlyEol]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:9:5: 'if rcurly' has incorrect indentation level 4, expected level should be 6. [Indentation]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:12:5: 'for lcurly' has incorrect indentation level 4, expected level should be 6. [Indentation]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:12:5: '{' at column 5 should be on the previous line. [LeftCurlyEol]
[WARN] /mnt/5D92528E6B945467/test/testing/Test.java:15:5: 'for rcurly' has incorrect indentation level 4, expected level should be 6. [Indentation]
Audit done.
```

Google-Formatter is not in favour of this type of formatting:

```java
$ java -jar google-java-format-1.27.0-all-deps.jar Test.java > FormattedCode.java
$ diff -Naru Test.java FormattedCode.java

--- Test.java   2025-08-06 18:26:43.215008200 +0530
+++ FormattedCode.java  2025-08-06 18:34:22.170034600 +0530
@@ -2,21 +2,19 @@
 public class Test {
 
   void test1(int[] nums) {
-    if (nums.length > 2)
-    {
+    if (nums.length > 2) {
       int temp = nums[0];
       use(temp);
     }
 
-    for (int i = 1; i < nums.length; i++) 
-    {
+    for (int i = 1; i < nums.length; i++) {
       int k = i * 2;
       use(k);
     }
   }
 
   void test2(int k) {
-    {  // ok
+    { // ok
       int x = k + 2;
       use(x);
     }
```

**Comment by Zopsss** [comment]
```java
 if (nums.length > 2)
    { // false-positive
      int temp = nums[0];
      use(temp);
    } // false-positive

    for (int i = 1; i < nums.length; i++) 
    {  // false-positive
      int k = i * 2;
      use(k);
    } // false-positive
```

how are these false-positive? the `{` should be on same line as if statement and for loop. The violation is correct

**Comment by mohitsatr** [comment]
@Zopsss 
when `{` appears on the same line as `if` statement or `for`, then it's in K&R block style.
However, this issue is about Exception to K&R block style i.e when `{` can be allowed on new line.

> Exception: In places where these rules allow a single statement ending with a semicolon (;), a block of statements can appear, and the opening brace of this block is preceded by a line break. Blocks like these are typically introduced to limit the scope of local variables.

**Comment by Zopsss** [comment]
this still doesn't explain why the examples I mentioned are marked as false-positive. 

---

> Blocks like these are typically introduced to limit the scope of local variables.

Here blocks means following type of block:

```java
    {
      int x = foo();
      frob(x);
    }
```

**Comment by mohitsatr** [comment]
> Braces follow the Kernighan and Ritchie style ("[Egyptian brackets](https://blog.codinghorror.com/new-programming-jargon/#3)") for nonempty blocks and block-like constructs:

```
for (...) {

}
```

```
if (..) {

}
```
aren't these `nonempty blocks` ?

one more point: 
> Line break after the closing brace, only if that brace terminates a statement or terminates the body of a method, constructor, or named class. For example, there is no line break after the brace if it is followed by `else` or a comma.

notice there is an `else` in the example, so `if`-`else` block must also be nonemptyblock. did I miss something?

**Comment by Zopsss** [comment]
> aren't these nonempty blocks ?

The examples you gave in issue description are non-empty because they contain at-least one statement. So the `if` and `for` statements' opening `{` should be on same line as the statement that starts the block. So `'{' at column 5 should be on the previous line.` violation is not a false-positive. It is saying to move `{` to it's previous line, where the `if` and `for` starts. 

A false-positive violation means an unnecessary violation, in other words a extra violation. The code follows the rule correctly but if Checkstyle still gives violation for it then it is considered as a false-positive violation.

Here in our case, we expect the `'{' at column 5 should be on the previous line.` violation. It is not an extra violation, so it is not a false-positive. 


If you're talking about the indentation violation, I mean if you're saying that the indentation violation is a false-positive then it is also wrong. Because we're not correctly following the style guide, if we follow the style guide correctly then both of these violations will go away. If we're not following style guide and gets an extra violation, but that violation goes away when we follow the style guide then I don't think it is a big deal. We don't need to fix that false-positive, fixing it also may affect may introduce some unwanted behavior of Indentation Check so we should avoid it. It will just waste our time, because user can fix that extra violation by following the style guide properly, it is not a big deal.

**Comment by Zopsss** [comment]
> notice there is an else in the example, so if-else block must also be nonemptyblock. did I miss something?

Yeah in some way, if-else block can be considered as nonemtpy block. An empty block should follow following style:

```java
if (something) {}
for (something) {}
```

The opening & closing braces should be on same line. But in case of if-else statement, it will look weird:

```java
if (something) {} else {}
```

so it should follow K&R styling regardless of it is actually empty or not:

```java
if (something) {
} else {
}
```

**Comment by romani** [comment]
>Exception: In places where these rules allow a single statement ending with a semicolon (;), a block of statements can appear, and the opening brace of this block is preceded by a line break. Blocks like these are typically introduced to limit the scope of local variables.

this is only for blocks methods like:
```
    {
      int x = foo();
      frob(x);
    }
```
whole method:
```
void foo() {
    int x = 6;
    {
      int x = boo();
      frob(x);
    }
   x++;
}
```


as in example:
```
  void test2(int k) {
    {  // ok
      int x = k + 2;
      use(x);
    }
  }
```
no violation , it is good.
Looks like this issue is false alert.
We can close it and remove reference to it from coverage  table .

**Comment by Zopsss** [comment]
Issue should be closed after merging https://github.com/checkstyle/checkstyle/pull/17651
